### PR TITLE
fix memory leak from exceptions

### LINKED ISSUE #436: memory leak with curry
We use dask for parallel computing, which in turn uses pytoolz and cytoolz.
We have a memory leak due to a reference not being dropped by dask. 

We wrote a tool to track down who is holding the reference to the memory.
We traced it back to the fact that when toolz inspects the signature of a function, it might return (not raise) an exception. The exception holds the frames and variables thus variables cannot be derenced until the exception is cleared. In #57, the result of inspecting a function is memoized, even if it's an exception. Thus, a memory leak is possible.

Our tool is gui based, so if you have a suggestion for where to upload images, I will be more than happy to provide the proof.

### PR DESCRIPTION
**Pull request by thejinx0r**
fix #436 
Not sure if there's a cleaner way, but this works.

### CODE REVIEW

**Comment by eriknw** [comment]
Thanks!  Looks reasonable to me.  We may want to look more closely and add a couple tests for when `sigspec` (as it's called in the code) is passed to other functions while it's None.

**Comment by thejinx0r** [comment]
@eriknw 
Since it's such a simple fix, and that you know the codebase much more than I do, can I leave you to write the tests for these?

From what I had seen, there shouldn't be any issue, but I wouldn't know what to test beyond the existing tests.

File: toolz/_signatures.py [hunk 1]
```diff
@@ -642,7 +642,7 @@ def signature_or_spec(func):
         try:
             return inspect.signature(func)
         except (ValueError, TypeError) as e:
```
Comment (akhilman) [inline toolz/_signatures.py]: This patch makes `e` unused variable.
```suggestion
        except (ValueError, TypeError):
```

File: toolz/_signatures.py [hunk 2]
```diff
@@ -664,7 +664,7 @@ def signature_or_spec(func):
         try:
             return inspect.getargspec(func)
         except TypeError as e:
```
Comment (akhilman) [inline toolz/_signatures.py]: This patch makes `e` unused variable.
```suggestion
        except TypeError:
```

**Comment by eriknw** [comment]
LGTM, thanks @thejinx0r and @akhilman!  We're past due for a release, so I'll get this up soon.
